### DISCUSSION TITLE
How do Drama's Actors compare to Sparks?

### DISCUSSION THREAD

**Discussion by dogatkeyboard** [evanrelf/drama#10] [Q&A] [answered]
1. How do Drama actors compare to Erlang actors?
2. Can Drama actors communicate across CPUs, or run in parallel (SMP enabled)?
3. Are Drama actors lighter weight than sparks?
4. How many Drama actors can run within a spark?
5. Can Drama actors run in the GPU?
6. What types of computation can be done with a Drama actor?

Included an image of reference for sparks:
https://i.stack.imgur.com/f57Hm.png from https://stackoverflow.com/questions/958449/what-is-a-spark-in-haskell

**Comment by dogatkeyboard** [comment]
There may be some redundancy in what I asked -- sorry if so. I think this is a wonderful project and just tried to pose some initial questions that came to me.

**Comment by evanrelf** [comment, marked as answer]
Hi @dogatkeyboard!

In general, Drama is more of an experiment to bring the actor model to Haskell. It's limited by what the GHC RTS supports, so it's nowhere near as sophisticated as what Erlang achieves on the BEAM virtual machine. I have not released a version 1.0.0.0 because I don't think it should be used in production.

To answer your questions specifically:

> 1. How do Drama actors compare to Erlang actors?

Drama actors are just GHC green threads with channels for communication. They are not separate OS processes or threads, and there is nothing preventing them from communicating or sharing memory outside of the API Drama provides.

> 2. Can Drama actors communicate across CPUs, or run in parallel (SMP enabled)?

Because Drama actors are just GHC green threads, you must run programs written with Drama with GHC's threaded runtime.

> 3. Are Drama actors lighter weight than sparks?

I'd actually never heard of sparks before. Drama actors are green threads, so they're heavier than sparks according to the Stack Overflow post you referenced.

> 4. How many Drama actors can run within a spark?

See answer to question 4.

> 5. Can Drama actors run in the GPU?

No.

> 6. What types of computation can be done with a Drama actor?

Literally anything you can do in Haskell. The `Actor` monad is just a `ReaderT` over `IO`.

**Reply by evanrelf** [reply]
@dogatkeyboard If you have any other questions, let me know. But I think you'll find that the Drama codebase is small and very readable, so reading the source should answer all of your questions.
